boards/shared/Src/Io/Io_SharedAdc.c

```diff
@@ -1,169 +1,140 @@
-static uint16_t GetVrefintCalibrationValue(void);
+// Pointer to the VREFINT measured by the ADC
+static uint16_t *measured_vrefint_address;
 
 /**
- * @brief Check if the ADC module has been initialized using
- *        Io_SharedAdc_StartAdcInDmaMode()
- * @note  This should be called at the beginning of functions that require ADC
- *        to be initialized
+ * Gets the ADC resolution for the given ADC handle
+ * @param hadc The ADC handle for the ADC we are initializing
+ * @return adc_max_resolution The maximum ADC resolution for the given ADC
+ * handle
  */
-static void CheckAdcInitialized(void);
+static ErrorStatus
+    Io_GetAdcResolution(ADC_HandleTypeDef *hadc, uint16_t *adc_max_resolution);
 
-static ErrorStatus InitializeAdcMaxValue(ADC_HandleTypeDef *hadc)
+static ErrorStatus
+    Io_GetAdcResolution(ADC_HandleTypeDef *hadc, uint16_t *adc_max_resolution)
 {
     switch (hadc->Init.Resolution)
     {
         case ADC_RESOLUTION_6B:
-            adc_max_value = MAX_6_BITS_VALUE;
+            *adc_max_resolution = MAX_6_BITS_VALUE;
             break;
         case ADC_RESOLUTION_8B:
-            adc_max_value = MAX_8_BITS_VALUE;
+            *adc_max_resolution = MAX_8_BITS_VALUE;
             break;
         case ADC_RESOLUTION_10B:
-            adc_max_value = MAX_10_BITS_VALUE;
+            *adc_max_resolution = MAX_10_BITS_VALUE;
             break;
         case ADC_RESOLUTION_12B:
-            adc_max_value = MAX_12_BITS_VALUE;
+            *adc_max_resolution = MAX_12_BITS_VALUE;
             break;
         default:
             return ERROR;
     }
+
     return SUCCESS;
 }
 
-static uint16_t GetVrefintCalibrationValue(void)
+struct AdcInput *
+    Io_SharedAdc_Init(ADC_HandleTypeDef *hadc, size_t vrefint_index)
 {
-    return (uint16_t)(*VREFINT_CALIBRATION_ADDRESS);
-}
+    assert(hadc != NULL);
 
-static void CheckAdcInitialized(void)
-{
-    if (adc_initialized != true)
-    {
-        Error_Handler();
-    }
-}
+    static size_t          alloc_index = 0;
+    static struct AdcInput adc_inputs[MAX_NUM_OF_ADC];
+    static uint32_t adc_readings[MAX_NUM_OF_ADC][MAX_NUM_OF_ADC_CHANNELS];
 
-void Io_SharedAdc_StartAdcInDmaMode(
-    ADC_HandleTypeDef *hadc,
-    uint32_t           vrefint_regular_rank)
-{
-    // The index of each ADC channel is always its Regular Rank minus one
-    vrefint_index = vrefint_regular_rank - 1;
-    bool status   = true;
+    struct AdcInput *const adc_input = &adc_inputs[alloc_index];
 
-    if (HAL_ADCEx_Calibration_Start(hadc, ADC_SINGLE_ENDED) != HAL_OK)
-    {
-        status = false;
-        Error_Handler();
-    }
+    assert(alloc_index < MAX_NUM_OF_ADC);
+    assert(Io_GetAdcResolution(hadc, &adc_input->adc_max_value) == SUCCESS);
 
-    if (HAL_ADC_Start_DMA(hadc, &adc_values[0], NUM_ADC_CHANNELS) != HAL_OK)
-    {
-        status = false;
-        Error_Handler();
-    }
+    adc_input->hadc                = hadc;
+    adc_input->num_active_channels = hadc->Init.NbrOfConversion;
```

Comment:
Why is it called active channels? I can't find it in the documentation and it seems misleading. I feel like this variable is redundant anyway because you're already saving a copy of hadc. If you need `NbrOfConversion` then just write `adc_iput->hadc->Init.NbrOfConversion`

Reply:
I feel like this also falls apart if you do over-sampling. Ie. if you sample each channel four times in a conversion sequence then your `hadc->Init.NbrOfConversion` will be _4*NUM_OF_CHANNELS_

Reply:
Yeah I was thinking about this too. I was thinking if there's a way to tweak the PR to accomodate for this. But as it stands, this would basically assume no oversampling of any kind and we always store only the latest copy of the ADC reading.

Reply:
For now, let's assume no oversampling in this PR. After everything is cleaned up let's revisit whether it's worth to making a shared library that assumes no oversampling.